### DISCUSSION TITLE
Progressive Texture resolutions loading

### DISCUSSION THREAD

**Discussion by lordcraymen** [google/model-viewer#4344] [Q&A]
I would like to load textures in power-of-two steps for two reasons:

1: UX - faster display of colors on the model.
2: Deliver only texture sizes that the device can handle. My full-resolution textures are 4096 in size. However, some devices can only handle sizes up to 2048 or even smaller. I check the possible texture size with gl.getParameter(gl.MAX_TEXTURE_SIZE).

I have a working backend service that creates and calculates the necessary solutions when I pass /url/originalfilename[_resolution].extension.

My original idea was to override the mapUrls function, which kind of works for the lowest resolution. However, I realize that my approach is probably too naive. When I use createTexture to load higher texture resolutions, the texture URI is passed again through mapUrls, rewriting it back to the URL of the lowest resolution.

There is likely a better or more appropriate way of achieving this. After several hours of searching online, I couldn't find the right solution.

Thank you for any pointers in the right direction!

```
const ModelViewerElement = customElements.get('model-viewer');
ModelViewerElement.mapURLs((url) => {
if (url.toLowerCase().includes(".jpg") || url.toLowerCase().includes(".png") ) {
    console.log("loading new texture",url);
    return url.replace(/(\.jpg|\.png|)$/i, "_128w$&");;
  }
  return url;
}
```
```
let baseColorTexture = modelviewer.model.material.pbrMetallicRoughness.baseColorTexture
viewer.createTexture("originalfilename.extention").then(t => { 
          console.log("loaded"+ newfilename, t);
          baseColorTexture.setTexture(t);
        } );
}
```

**Comment by elalish** [comment]
You might just author a GLB with the lowest-res textures to get the fastest load, then pull in higher res using `createTexture` as appropriate, so you can skip `mapUrls` entirely. 

I'd also recommend considering the pixel resolution that MV is displayed at maximum allowed zoom to set your texture sizes, rather than always using the max the device can handle. This will help with frame rate and battery draw.

**Reply by lordcraymen** [reply]
Thank you for your suggestion regarding the distance/zoom of the camera, I’ll try that. For the „lowest“ possible texture size I am not convinced. I know what the target device can handle at runtime, so I think it’s best to figure out the appropriate solution at runtime as well. 

Is there an event specifically if a texture is loaded or applied? I would prefer not to loop through all materials to reapply higher versions of the textures, it would be nicer if I could listen for the successful loading of a texture to start loading the next higher resolution. Or is it possible to hook into the texture loading functionality of the model-viewer itself, like with the static useMaps function? Thank you for your support!

**Reply by elalish** [reply]
All I can think of is load a GLB with no textures at all, and concurrently do `createTexture`. Then you know when they come down because the promise resolves. It won't be any slower than having them packed in the GLB.
